src/control/server/server_utils.go

```diff
@@ -0,0 +1,356 @@
+	if cfg.HelperLogFile != "" {
+		if err := setenv(pbin.DaosAdminLogFileEnvVar, cfg.HelperLogFile); err != nil {
+			return errors.Wrap(err, "unable to configure privileged helper logging")
+		}
+	}
+
+	if cfg.FWHelperLogFile != "" {
+		if err := setenv(pbin.DaosFWLogFileEnvVar, cfg.FWHelperLogFile); err != nil {
+			return errors.Wrap(err, "unable to configure privileged firmware helper logging")
+		}
+	}
+
+	return nil
+}
+
+// Provide special handling for the ofi+verbs provide.
+// Mercury uses the interface name such as ib0, while OFI uses the device name
+// such as hfi1_0 CaRT and Mercury will now support the new OFI_DOMAIN
+// environment variable so that we can specify the correct device for each.
+func updateFabricEnvars(ctx context.Context, ec *engine.Config) error {
+	if strings.HasPrefix(ec.Fabric.Provider, "ofi+verbs") && !ec.HasEnvVar("OFI_DOMAIN") {
+		deviceAlias, err := netdetect.GetDeviceAlias(ctx, ec.Fabric.Interface)
+		if err != nil {
+			return errors.Wrapf(err, "failed to resolve alias for %s", ec.Fabric.Interface)
+		}
+		envVar := "OFI_DOMAIN=" + deviceAlias
+		ec.WithEnvVars(envVar)
+	}
+
+	return nil
+}
+
+func registerEngineCallbacks(engine *EngineInstance, pubSub *events.PubSub, allStarted *sync.WaitGroup) {
+	// Register callback to publish I/O Engine process exit events.
+	engine.OnInstanceExit(publishInstanceExitFn(pubSub.Publish, hostname(), engine.Index()))
+
+	var onceReady sync.Once
+	engine.OnReady(func(_ context.Context) (err error) {
+		// Indicate that engine has been started, only do this
+		// the first time that the engine starts as shared
+		// memory persists between engine restarts.
+		onceReady.Do(func() {
+			allStarted.Done()
+		})
+		return
+	})
+}
+
+func initFirstEngine(engine *EngineInstance, sysdb *system.Database, joinFn systemJoinFn) error {
+	if !sysdb.IsReplica() {
+		return nil
+	}
+
+	// Start the system db after instance 0's SCM is
+	// ready.
+	var onceStorageReady sync.Once
+	engine.OnStorageReady(func(ctx context.Context) (err error) {
+		onceStorageReady.Do(func() {
+			err = errors.Wrap(sysdb.Start(ctx),
+				"failed to start system db",
+			)
+		})
+		return
+	})
+
+	if !sysdb.IsBootstrap() {
+		return nil
+	}
+
+	// For historical reasons, we reserve rank 0 for the first
+	// instance on the raft bootstrap server. This implies that
+	// rank 0 will always be associated with a MS replica, but
+	// it is not guaranteed to always be the leader.
+	engine.joinSystem = func(ctx context.Context, req *control.SystemJoinReq) (*control.SystemJoinResp, error) {
+		if sb := engine.getSuperblock(); !sb.ValidRank {
+			engine.log.Debug("marking bootstrap instance as rank 0")
+			req.Rank = 0
+			sb.Rank = system.NewRankPtr(0)
+		}
+		return joinFn(ctx, req)
+	}
+
+	return nil
+}
+
+func exportStats(ctx context.Context, allStarted *sync.WaitGroup, log *logging.LeveledLogger, port int, instances []*EngineInstance) {
```

Comment:
It might be worth reworking this a little bit to call exportStats in a closure that is added to a list of closures to be called when `allStarted.Wait()` unblocks. That way we can easily add other logic that needs to wait for all of the engines to be started before running.